src/internal/pfsdb/commits.go

```diff
@@ -485,6 +489,56 @@ func getCommitChildren(ctx context.Context, extCtx sqlx.ExtContext, parentCommit
 	return children, nil
 }
 
+// GetCommitAncestry returns a CommitAncestry from startId up to maxDepth.
+func GetCommitAncestry(ctx context.Context, extCtx sqlx.ExtContext, startId CommitID, maxDepth uint) (map[CommitID]CommitID, error) {
+	ancestry := make(map[CommitID]CommitID)
+	if err := ForEachCommitAncestor(ctx, extCtx, startId, maxDepth, func(parentId, childId CommitID) error {
+		ancestry[parentId] = childId
+		return nil
+	}); err != nil {
+		return nil, errors.Wrap(err, "get commit ancestry")
+	}
+	return ancestry, nil
+}
+
+// ForEachCommitAncestor queries postgres for ancestors of startId up to the maxDepth. cb() is called for each ancestor.
+// maxDepth is optional.
+func ForEachCommitAncestor(ctx context.Context, extCtx sqlx.ExtContext, startId CommitID, maxDepth uint, cb func(parentId, childId CommitID) error) error {
+	if maxDepth == 0 {
+		maxDepth = DefaultMaxSearchDepth
+	}
+	query := `
+	WITH RECURSIVE ancestry AS (
+		SELECT parent, child, 1 as depth FROM pfs.commit_ancestry WHERE child = $1
+		UNION
+		SELECT ca.parent, ca.child, depth+1 FROM pfs.commit_ancestry ca
+		JOIN ancestry a ON ca.child = a.parent WHERE depth <= $2
+	)
+	SELECT a.parent, a.child, depth
+	FROM ancestry a WHERE depth <= $2;`
```

Comment:
If the recursion above is correct, then the where clause shouldn't be needed here. Also, I think the where clause above might need to be `<` rather than `<=` because it applies to the current state of the ancestry table.